### PR TITLE
support for jakarta and javax violations maven enforcer rule

### PR DESCRIPTION
**Pull request by trentjeff**
This PR offers a new maven enforcer plugin rule that will look at all compile and runtime transitive dependencies of a given module for any modules that are unsupported by Helidon.

It can be invoked as follows:
```xml
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-enforcer-plugin</artifactId>
    <dependencies>
        <dependency>
            <groupId>io.helidon.build-tools</groupId>
            <artifactId>helidon-build-maven-enforcer-rules</artifactId>
            <version>4.0.0-SNAPSHOT</version>
        </dependency>
    </dependencies>
    <executions>
        <execution>
            <id>enforce-dependencies</id>
            <goals>
                <goal>enforce</goal>
            </goals>
            <configuration>
                <rules>
                    <helidonDependenciesRule>
                        <namespace>JAKARTA | JAVAX</namespace>
                        <excludedGavRegExs/>
                    </helidonDependenciesRule>
                </rules>
            </configuration>
        </execution>
    </executions>
</plugin>
```

The integration into the main build is covered by https://github.com/helidon-io/helidon/issues/7513

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -263,7 +264,7 @@
                 <plugin>
                     <groupId>org.apache.maven.plugins</groupId>
                     <artifactId>maven-enforcer-plugin</artifactId>
-                    <version>${version.plugin.enforcer}</version>
+                    <version>${version.plugin.maven-enforcer}</version>
```
Comment (arjav-desai) [inline pom.xml:267]: Seems the property name above in line 187 is missing "maven"

Comment (trentjeff) [reply pom.xml:267]: That is the way it was already defined in the top-level pom under the banner "        
"

File: common/maven-enforcer-rules/pom.xml [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            <version>${version.lib.shrinkwrap}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.maven</groupId>
+            <artifactId>maven-artifact</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.maven.enforcer</groupId>
+            <artifactId>enforcer-api</artifactId>
+            <version>${version.lib.maven-enforcer}</version>
```
Comment (arjav-desai) [inline common/maven-enforcer-rules/pom.xml:55]: nit pick: shouldn't be needed as its coming from parent pom.

File: pom.xml (hunk 1 above)
Comment (arjav-desai) [reply pom.xml:268]: What I am saying is, you have renamed properties for "lib" to add maven at both places (property as well as usage) but for "plugin" you have just added it here in usage but definition of it is still without maven on line 187 i.e. ${version.plugin.enforcer}

Comment (trentjeff) [reply pom.xml]: pushed

File: common/maven-enforcer-rules/src/main/java/io/helidon/build/common/maven/enforcer/rules/HelidonJakartaDependenciesRule.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    //    @Inject
+    private List<String> excludedGavs;
+
+    @Override
+    public void execute() throws EnforcerRuleException {
+        ConfigurableMavenResolverSystem resolver = Maven.configureResolver()
+                .workOffline()
+                .withClassPathResolution(false);
+        List<Gav> deps = resolver
```
Comment (romain-grecourt) [inline common/maven-enforcer-rules/src/main/java/io/helidon/build/common/maven/enforcer/rules/HelidonJakartaDependenciesRule.java:64]: This is a simple use-case that should not require a third party. I.e. use "maven-resolver".

File: pom.xml [hunk 4]
```diff
@@ -966,6 +973,7 @@
                     <plugin>
                         <groupId>io.helidon.build-tools</groupId>
                         <artifactId>helidon-enforcer-plugin</artifactId>
+                        <!-- TODO: should align to 3.4.0 here? -->
```
Comment (trentjeff) [inline pom.xml:976]: Romain?

File: maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/HelidonDependenciesRule.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    @Override
+    public void execute() throws ViolationException {
+        String namespace = checkNamespace(this.namespace);
+        List<Pattern> excludedGavRegExs = this.excludedGavRegExs.stream()
+                .map(Pattern::compile)
+                .collect(Collectors.toList());
+        List<Artifact> artifacts = project.getArtifacts().stream()
+                .filter(a -> a.getGroupId().startsWith("javax.") || a.getGroupId().startsWith("jakarta."))
+                .filter(a -> a.getScope().equalsIgnoreCase("compile") || a.getScope().equalsIgnoreCase("runtime"))
```
Comment (barchetta) [inline maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/HelidonDependenciesRule.java:77]: I think this should also filter out optional dependencies.

File: common/maven-enforcer-rules/src/main/java/io/helidon/build/common/maven/enforcer/rules/HelidonJakartaDependenciesRule.java (hunk 3 above)
Comment (trentjeff) [reply common/maven-enforcer-rules/src/main/java/io/helidon/build/common/maven/enforcer/rules/HelidonJakartaDependenciesRule.java]: pushed

File: maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/HelidonDependenciesRule.java (hunk 5 above)
Comment (trentjeff) [reply maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/HelidonDependenciesRule.java:77]: pushed

File: pom.xml (hunk 4 above)
Comment (romain-grecourt) [reply pom.xml:976]: I don't understand what you are asking. This is our own "enforcer" plugin and since it's housed in this repository we have to depend on a previously released version.

File: maven-enforcer-rules/README.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+            <executions>
+                <execution>
+                    <id>enforce-helidon-dependencies</id>
+                    <goals>
+                        <goal>enforce</goal>
+                    </goals>
+                    <configuration>
+                        <rules>
+                            <helidonDependenciesRule>
+                                <namespace>JARKTA | JAVAX</namespace>
```
Comment (romain-grecourt) [inline maven-enforcer-rules/README.md:34]: `JARKTA` -> `JAKARTA`, also this should probably be documented as lowercase, even if the string is case insensitive.

File: maven-enforcer-rules/pom.xml [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>generate-metadata</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
```
Comment (romain-grecourt) [inline maven-enforcer-rules/pom.xml:85]: Why is this needed ?

File: maven-enforcer-rules/pom.xml [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.maven</groupId>
+            <artifactId>maven-artifact</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.maven.enforcer</groupId>
+            <artifactId>enforcer-api</artifactId>
+            <version>${version.lib.maven-enforcer}</version>
```
Comment (romain-grecourt) [inline maven-enforcer-rules/pom.xml:47]: Can you add dependency management for this dependency in the top-level parent pom ?

File: maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheck.java [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+            return true;
+        }
+
+        if (groupPackageName.equals("javax.servlet")
+                || groupPackageName.equals("jakarta.servlet")) {
+            return false;
+        }
+
+        try {
+            if (HelidonDependenciesRule.JAKARTA.equals(namespace)) {
```
Comment (romain-grecourt) [inline maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheck.java:74]: `equalsIgnoreCase`

File: maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheck.java [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+                return applyJakartaRule(groupPackageName, gav.getSelectedVersion());
+            } else if (HelidonDependenciesRule.JAVAX.equals(namespace)) {
```
Comment (romain-grecourt) [inline maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheck.java:76]: `equalsIgnoreCase`

File: maven-enforcer-rules/src/test/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheckTest.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.startsWith;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class DependencyIsValidCheckTest {
```
Comment (romain-grecourt) [inline maven-enforcer-rules/src/test/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheckTest.java:31]: Tests are usually package private.

File: maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/HelidonDependenciesRule.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    @Inject
+    private MavenProject project;
+
+    @Inject
+    private MavenSession session;
+
+    /**
+     * Rule parameter as list of gav regular expressions. Reflectively injected from pom.xml
+     */
+    // @Inject
```
Comment (romain-grecourt) [inline maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/HelidonDependenciesRule.java:57]: Commented code

File: pom.xml (hunk 4 above)
Comment (trentjeff) [reply pom.xml]: oops, pushed

File: maven-enforcer-rules/pom.xml (hunk 7 above)
Comment (trentjeff) [reply maven-enforcer-rules/pom.xml:85]: this was a leftover from when I was supporting a Main.  I sinced removed that so will remove that now.

File: maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheck.java (hunk 9 above)
Comment (trentjeff) [reply maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheck.java:74]: technically that is not needed here, but will add it for extra robustness.

File: maven-enforcer-rules/src/test/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheckTest.java (hunk 11 above)
Comment (trentjeff) [reply maven-enforcer-rules/src/test/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheckTest.java:31]: good catch - next push

File: maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/HelidonDependenciesRule.java (hunk 12 above)
Comment (trentjeff) [reply maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/HelidonDependenciesRule.java:57]: intended to relay that this is injected by maven.  i plan to keep this that way.

File: maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/HelidonDependenciesRule.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    private List<String> excludedGavRegExs;
+
+    /**
+     * Rule parameter string. The enforcement policy can either be {@link #JAVAX} or {@link #JAKARTA}.
+     * Reflectively injected from pom.xml. The default is {@link #JAKARTA}.
+     * When set to {@link #JAKARTA}, only "post javax->jakarta" module names are allowed. When set to {@link #JAVAX}, only
+     * "pre javax->jakarta" module names are allowed.
+     */
+    // @Inject
```
[the remark this answers is not in the document]
Comment (trentjeff) [reply maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/HelidonDependenciesRule.java:66]: same here

File: maven-enforcer-rules/src/test/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheckTest.java (hunk 11 above)
Comment (trentjeff) [reply maven-enforcer-rules/src/test/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheckTest.java]: pushed

File: maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheck.java (hunk 9 above)
Comment (trentjeff) [reply maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/DependencyIsValidCheck.java]: pushed

File: maven-enforcer-rules/README.md (hunk 6 above)
Comment (trentjeff) [reply maven-enforcer-rules/README.md]: fixed

File: maven-enforcer-rules/pom.xml (hunk 8 above)
Comment (trentjeff) [reply maven-enforcer-rules/pom.xml]: done and pushed

File: maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/HelidonDependenciesRule.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (trentjeff) [reply maven-enforcer-rules/src/main/java/io/helidon/build/maven/enforcer/rules/HelidonDependenciesRule.java]: pushed

**Review by romain-grecourt — APPROVED** [review]
